src/mobu/main.py

```diff
@@ -9,47 +9,117 @@
 
 from __future__ import annotations
 
+import json
 from collections.abc import AsyncIterator
-from contextlib import asynccontextmanager
+from contextlib import AsyncExitStack, asynccontextmanager
 from datetime import timedelta
 from importlib.metadata import metadata, version
 
 import structlog
 from fastapi import FastAPI
+from fastapi.openapi.utils import get_openapi
 from safir.fastapi import ClientRequestError, client_request_error_handler
 from safir.logging import Profile, configure_logging, configure_uvicorn_logging
 from safir.middleware.x_forwarded import XForwardedMiddleware
 from safir.slack.webhook import SlackRouteErrorHandler
 
 from .asyncio import schedule_periodic
 from .config import config
-from .dependencies.context import context_dependency
+from .dependencies.context import ContextDependency, context_dependency
+from .dependencies.github import (
+    ci_manager_dependency,
+    github_config_dependency,
+)
 from .handlers.external import external_router
-from .handlers.github import github_router
+from .handlers.github_ci_app import api_router as github_ci_app_router
+from .handlers.github_refresh_app import (
+    api_router as github_refresh_app_router,
+)
 from .handlers.internal import internal_router
 from .status import post_status
 
 __all__ = ["app", "lifespan"]
 
 
 @asynccontextmanager
-async def lifespan(app: FastAPI) -> AsyncIterator[None]:
-    """Set up and tear down the the application."""
+async def base_lifespan(app: FastAPI) -> AsyncIterator[ContextDependency]:
+    """Set up and tear down the the base application."""
     if not config.environment_url:
         raise RuntimeError("MOBU_ENVIRONMENT_URL was not set")
     if not config.gafaelfawr_token:
         raise RuntimeError("MOBU_GAFAELFAWR_TOKEN was not set")
+
     await context_dependency.initialize()
     await context_dependency.process_context.manager.autostart()
+
     status_interval = timedelta(days=1)
     app.state.periodic_status = schedule_periodic(post_status, status_interval)
 
-    yield
+    yield context_dependency
 
     await context_dependency.aclose()
     app.state.periodic_status.cancel()
 
 
+@asynccontextmanager
+async def github_ci_app_lifespan(
+    base_context: ContextDependency,
+) -> AsyncIterator[None]:
+    """Set up and tear down the GitHub CI app functionality."""
+    if not config.github_config_path:
+        raise RuntimeError("MOBU_GITHUB_CONFIG_PATH was not set")
+    if not config.github_ci_app.webhook_secret:
+        raise RuntimeError("MOBU_GITHUB_CI_APP_WEBHOOK_SECRET was not set")
+    if not config.github_ci_app.private_key:
+        raise RuntimeError("MOBU_GITHUB_CI_APP_PRIVATE_KEY was not set")
+    if not config.github_ci_app.id:
+        raise RuntimeError("MOBU_GITHUB_CI_APP_ID was not set")
```

Comment:
I think it's best to leave this as future work, since this PR is already really long and we should get it merged to keep things digestible, but for the future I think you can simplify this initialization story a lot by moving a lot of these checks into validators in the config models. Getting the sequencing right with the test suite can be a little tricky, but usually you can just set the necessary environment variables in `tox.ini`.

Reply:
https://github.com/lsst-sqre/mobu/pull/354